### PR TITLE
Fake and real window fixtures for tests

### PR DESCRIPTION
**Pull request by dvoytenko**

### CODE REVIEW

File: testing/describes.js [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+ * })} fn
+ */
+export function realWin(name, spec, fn) {
+  return describe(name, function() {
+
+    const test = Object.create(null);
+
+    beforeEach(() => {
+      return new Promise(function(resolve, reject) {
+        const iframe = document.createElement('iframe');
```
Comment (cramforce) [inline testing/describes.js]: Can this use `iframe.js`? I suppose you did this on purpose?

File: testing/describes.js [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+ * @param {string} name
+ * @param {{
+ *   fakeClock: (boolean|undefined),
+ * }} spec
+ * @param {function({
+ *   sandbox: !Sandbox,
+ *   clock: (!FakeClock|undefined),
+ * })} fn
+ */
+export function sandboxed(name, spec, fn) {
```
Comment (lannka) [inline testing/describes.js]: ha, shall we just make this the default? any use case a sandbox can be unwanted?

File: testing/describes.js [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+ * }} spec
+ * @param {function({
+ *   sandbox: !Sandbox,
+ *   clock: (!FakeClock|undefined),
+ * })} fn
+ */
+export function realWin(name, spec, fn) {
+  return describe(name, function() {
+
+    const test = Object.create(null);
```
Comment (lannka) [inline testing/describes.js]: nitpick: how about call it `env`

File: test/functional/test-extensions.js [hunk 4]
```diff
@@ -25,52 +25,31 @@ import {
...
-} from '../../testing/iframe';
 import {
   initLogConstructor,
   resetLogConstructorForTesting,
 } from '../../src/log';
 import {loadPromise} from '../../src/event-helper';
-import * as cust from '../../src/custom-element';
-import * as sinon from 'sinon';
-
+import * as describes from '../../testing/describes';
```
Comment (erwinmombay) [inline test/functional/test-extensions.js]: should we make it a global instead?

File: test/functional/test-extensions.js [hunk 5]
```diff
@@ -25,52 +25,31 @@ import {
...
-    sandbox = sinon.sandbox.create();
+    sandbox = test.sandbox;
   });
 
   afterEach(() => {
     sandbox.restore();
   });
 
-  describe('registerExtension', () => {
+  describes.fakeWin('registerExtension', {}, test => {
```
Comment (lannka) [inline test/functional/test-extensions.js]: just curious why the tests have to be done in a fake window?

File: testing/describes.js [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+      ampDestroy(test);
+      // TODO(dvoytenko): test that window is returned in a good condition.
+      if (test.iframe.parentNode) {
+        test.iframe.parentNode.removeChild(test.iframe);
+      }
+      delete test.iframe;
+      delete test.win;
+    });
+
+    describe(SUB, function() {
```
Comment (lannka) [inline testing/describes.js]: Is the SUB reserved for future expansion to tests under more env?

File: testing/describes.js [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+      fn.call(this, test);
+    });
+  });
+}
+
+
+/**
+ * A test with a fake window.
```
Comment (erwinmombay) [inline testing/describes.js]: should it be real win?

File: testing/fake-dom.js [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+
+    /**
+     * @param {number} id
+     * @const
+     */
+    this.clearTimeout = window.clearTimeout;
+
+    /**
+     * @param {function()} handler
+     * @param {number=} timeout
```
Comment (erwinmombay) [inline testing/fake-dom.js]: missing the var_args on the sig

File: test/functional/test-extensions.js (hunk 4 above)
Comment (dvoytenko) [inline test/functional/test-extensions.js]: Done

File: testing/describes.js (hunk 2 above)
Comment (dvoytenko) [inline testing/describes.js]: I was thinking about that. But for migration I decided to do `sandboxed` instead. You still need to do top-level `describe` and we'll adopt `sandboxed` instead until all tests are migrated. After that, we can push it into all tests by default.

File: testing/describes.js (hunk 3 above)
Comment (dvoytenko) [inline testing/describes.js]: Done

File: testing/describes.js (hunk 6 above)
Comment (dvoytenko) [inline testing/describes.js]: No, but I have to call `describe` here to get tests-own beforeEach and afterEach called and if I call `describe`, it must have a name, even if it's just a space.

File: testing/fake-dom.js (hunk 8 above)
Comment (dvoytenko) [inline testing/fake-dom.js]: Done

File: test/functional/test-extensions.js (hunk 4 above)
Comment (dvoytenko) [inline test/functional/test-extensions.js]: I also made `sandbox` global. Just `sandbox` for now to see how it plays out. We may do this for other vars as well.

**Comment by dvoytenko** [comment]
All! PTAL, all comments addressed.

@lannka, I don't have a story on `postMessage` yet. But I'm hoping, once this is in, we will all fill in the gaps.

File: testing/describes.js [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+ * }} spec
+ * @param {function({
+ *   sandbox: !Sandbox,
+ *   clock: (!FakeClock|undefined),
+ * })} fn
+ */
+export function sandboxed(name, spec, fn) {
+  return describe(name, function() {
+
+    const env = Object.create(null);
```
Comment (erwinmombay) [inline testing/describes.js]: @dvoytenko should we also freeze or seal the env objet?

File: testing/fake-dom.js [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+ */
+class EventListeners {
+
+  /**
+   * @param {!EventTarget} target
+   * @return {!EventListeners}
+   */
+  static intercept(target) {
+    target.eventListeners = new EventListeners();
+    target.addEventListener = function(type, handler, captureOrOpts) {
```
Comment (erwinmombay) [inline testing/fake-dom.js]: awesome

**Comment by erwinmombay** [comment]
LGTM deferring for a 2nd LGTM

File: testing/describes.js (hunk 2 above)
Comment (lannka) [inline testing/describes.js]: How about just making it a part of the realWinSpec then?
I just don't like that we need to repeat such a hiarachy in every test:
```
describes.sandboxed(()=>{
  describes.readWindow()=>
```

**Comment by dvoytenko** [comment]
@lannka Hmm. This nesting is optional. But let me try to remove it altogether...

**Comment by dvoytenko** [comment]
@lannka I refactored things a bit. Now all describes create a sandbox (or reuse an existing one). This definitely looks better. PTAL.

File: testing/describes.js [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+        fixture.teardown(env);
+      });
+
+      // Delete all other keys.
+      for (const key in env) {
+        delete env[key];
+      }
+    });
+
+    describe(SUB, function() {
```
Comment (jridgewell) [inline testing/describes.js]: We do we have another `describe` block?

Comment (dvoytenko) [inline testing/describes.js]: So that the callback we call (`fn`) can call its own `beforeEach` and `afterEach` and we observe the nested behavior.

**Pull request merged**